---
title: "LTR Parser Results"
author: "Kevin McCormick"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  pdf_document: 
    keep_tex: no
    df_print: kable
header-includes:
- \usepackage{booktabs}
- \usepackage{longtable}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warnings = FALSE, message = FALSE, warning = FALSE, results = "asis")
options(knitr.table.format = "latex") 

```

# Overview

This report contains the results of attempted primer & LTR identification in the following FASTQ file:

`r inputFastq`

```{r summary}

summaryStats %>%
  gather("Variable", "Value") %>%
  mutate(Variable=gsub("\\."," ", Variable)) %>%
  kableExtra::kable(col.names = NULL)

```

\newpage

Visualization of `r min(1000, sum(seqsDf$count))` sequencing reads, selected randomly and arranged by nucleotide sequence:

```{r plot1}

dnaplotr::plotDNA(seqsDf %>%
                    sample_n(min(1000, sum(count)), replace = TRUE, weight = count) %>%
                    arrange(seq) %>%
                    pull(seq))

```

Visualization of `r min(1000, nrow(seqsDf))` distinct sequencing reads (after removing duplicates), selected randomly and arrangedy by nucleotide sequence:

```{r plot2}


dnaplotr::plotDNA(seqsDf %>%
                    sample_n(min(1000, n()), replace = FALSE) %>%
                    arrange(seq) %>%
                    pull(seq))

```

```{r}

if(nrow(commonCandidates) ==0) {
  cat("No primer LTR pairs appeared in more than 5% of total reads.")
  knitr::knit_exit()
}

```

\newpage

We found  `r nrow(outputTable)` potential primer-LTR pairs, `r nrow(outputTable %>% filter(LANL))` of which has an LTR sequence that mapped to an LTR region of at least one genome in the Los Alamos National Laboratories HIV database.  The "count" column here is the number of reads with each LTR-primer pair that also mapped to the human genome.


```{r ltrs}
outputTable %>%
  select(-MD, -qname) %>%
  kableExtra::kable(longtable = TRUE) %>%
  kableExtra::kable_styling(font_size = 7, bootstrap_options = c("striped", "condensed"))
```

Visualization of `r min(1000, sum(filteredSeqsDf$count))` sequencing reads filtered for common primer & LTR, selected randomly and arranged by nucleotide sequence (these reads may or may not map to the human genome):

```{r ltrViz}
dnaplotr::plotDNA(filteredSeqsDf %>%
                    sample_n(min(1000, sum(count)), replace = TRUE, weight = count) %>%
                    arrange(seq) %>%
                    pull(seq))

```
